### PR TITLE
Metric logger improvements

### PR DESCRIPTION
**Pull request by ebsmothers**
This is inspired by @tcapelle's changes in #730 so props to him for getting the ball rolling here. 

Quick primer/definition to make what follows clearer: by **iteration** I mean a single batch from our dataloader. By **step** I mean a single step with our optimizer. So without gradient accumulation `iteration == step`, with gradient accumulation there are `gradient_accumulation_steps` iterations per step (maybe the naming of this field is itself. confusing tbh)

The main set of changes here is:

- Actually log once on every step rather than on every iteration of a step (as is being done [here](https://github.com/pytorch/torchtune/blob/a46560ea428939a8f5d91c7b49f189ff1787da28/recipes/full_finetune_single_device.py#L412). This only matters when using gradient accumulation, but that's actually done for a lot of default configs so we need to get it right.
- Accumulate loss correctly over iterations for logging. E.g. instead of just taking the loss on the final iteration, we sum it up over each iteration (normalized by `gradient_accumulation_steps`).
- Add `tokens_per_second` metric. Similarly this is accumulated over iterations to log the per-step average.
- Add a config field to control whether we log peak memory or not. Currently this was just a hardcoded `log_peak_memory_every_n_steps=100` field, which (a) was completely arbitrary and (b) bloated up our training loop with multiple `log_dict` calls and multiple different logging frequencies. For BC with existing configs I do a default check in the recipe and set it to False
- Fix our pbar to also be per-step instead of per-iteration. Otherwise it doesn't line up with our metric logger which is confusing
- Skip our wandb logger test (for now). It was never actually running and it's broken 😢 

### Full finetune single device
```
tune run full_finetune_single_device --config llama3/8B_full_single_device metric_logger=torchtune.utils.metric_logging.WandBLogger metric_logger.project=lora-debug optimizer_in_bwd=False gradient_accumulation_steps=2 log_every_n_steps=1 log_peak_memory_stats=True max_steps_per_epoch=10 
epochs=1
...
1|10|Loss: 0.9859869480133057: 100%|█████████████| 10/10 [00:15<00:00,  1.14s/it]
```
[Metric logger outputs](https://wandb.ai/ebs/lora-debug/runs/kalhm3z3)

### Full finetune distributed
```
tune run --nproc_per_node 2 full_finetune_distributed --config gemma/2B_full metric_logger=t
orchtune.utils.metric_logging.WandBLogger metric_logger.project=lora-debug gradient_accumulation_steps=2 log_every_n_steps=1 log_peak_memory_stats=True max_steps_per_epoch=10 epochs=1
...
1|10|Loss: 1.1142131090164185: 100%|█████████| 10/10 [00:15<00:00,  1.50s/it]
```
[Metric logger outputs](https://wandb.ai/ebs/lora-debug/runs/dky74q35)

### LoRA finetune single device
```
tune run lora_finetune_single_device --config llama2/7B_lora_single_device metric_logger=torchtune.utils.metric_logging.WandBLogger metric_logger.project=lora-debug gradient_accumulation_steps=2 log_every_n_steps=1 log_peak_memory_stats=True max_steps_per_epoch=10 epochs=1
...
1|10|Loss: 1.6541345119476318: 100%|██████████████| 10/10 [00:07<00:00,  1.59it/s]
```
[Metric logger outputs](https://wandb.ai/ebs/lora-debug/runs/a8xekgjj)

### LoRA finetune distributed
```
tune run --nproc_per_node 2 lora_finetune_distributed --config mistral/7B_lora metric_logger=torchtune.utils.metric_logging.WandBLogger metric_logger.project=lora-debug gradient_accumulation_steps=2 log_every_n_steps=1 log_peak_memory_stats=True max_steps_per_epoch=10 epochs=1
...
1|5|Loss: 2.1742944717407227:  50%|█████                                                                 | 5/10 [00:16<00:16,  3.24s/it]
```
[Metric logger outputs](https://wandb.ai/ebs/lora-debug/runs/javinyzo)

### LoRA DPO single device
```
tune run lora_dpo_single_device --config llama2/7B_lora_dpo_single_device metric_logger=torchtune.utils.metric_logging.WandBLogger metric_logger.project=lora-debug gradient_accumulation_steps=2 log_every_n_steps=1 log_peak_memory_stats=True max_steps_per_epoch=10 epochs=1
...
1|7|Loss: 0.6841517686843872:  70%|█████████                                       | 7/10 [00:14<00:05,  1.91s/it]
```
[Metric logger outputs](https://wandb.ai/ebs/lora-debug/runs/7mxliits)

### LoRA DPO distributed
```
tune run --nproc_per_node 2 lora_dpo_distributed --config llama2/7B_lora_dpo metric_logger=torchtune.utils.
metric_logging.WandBLogger metric_logger.project=lora-debug gradient_accumulation_steps=2 log_every_n_steps=1 log_peak_memory_stats=True max_steps_per_epoch=10 epochs=1
...
1|8|Loss: 0.6825119256973267:  80%|████████                          | 8/10 [00:20<00:04,  2.42s/it]
```
[Metric logger outputs](https://wandb.ai/ebs/lora-debug/runs/w00j8bby)

### CODE REVIEW

**Comment by tcapelle** [comment]
Hey, can I help you here? Looks similiar to what I was working: https://github.com/pytorch/torchtune/pull/730

File: recipes/full_finetune_distributed.py [hunk 1]
```diff
@@ -468,44 +477,47 @@ def train(self) -> None:
...
+                    # Log per-step metrics
+                    if (
+                        self.total_training_steps % self._log_every_n_steps == 0
+                        and self._is_rank_zero
+                    ):
+                        time_per_step = time.perf_counter() - t0
+                        log_dict = {
+                            "loss": loss_to_log,
+                            "lr": self._optimizer.param_groups[0]["lr"],
+                            "tokens_per_second": num_tokens / time_per_step,
```
[the remark this answers is not in the document]
Comment (ebsmothers) [reply recipes/full_finetune_distributed.py:507]: Yeah this is a good point. Actually I'm not sure what the right thing to do here is for two reasons. (1) We only log rank-zero loss, and (2) I think a naive multiplication with num_devices will not work. If I understand correctly DistributedSampler will apply collate_fn to each worker's batch separately, in which case we actually have different numbers of tokens on each rank, right?

[the remark this answers is not in the document]
Comment (ebsmothers) [reply recipes/full_finetune_distributed.py:507]: Yeah the more I think about it the less I want to change this. (1) it will introduce unneeded syncs to calculate this value properly and (2) we could just log on every rank if we really want to and aggregate posthoc

**Review by RdoubleA — APPROVED** [review]
